### ISSUE TITLE
Handling statefulset recreation without user facing downtime

### ISSUE THREAD

**Issue by tiagomeireles** [prometheus-operator/prometheus-operator#3875]
How are others with large prometheus footprint handling statefulset recreation without user facing downtime?
Currently I run 3 prometheus pods scraping the same targets fronted by Thanos. The statefulset recreation causes all 3 prometheus pods to go offline. The prometheus pods take longer than our scrape interval to become available(5min+ in our case) which would result in a user facing downtime.

How does the prometheus-operator handle syncing prometheus objects? Does it sync objects serially or in parallel? 
One idea that I had was to switch to running 2 separate statefulsets with 2 replicas but that would be dependent on objects being synced serially.

**Comment by coderanger** [comment]
As I said on Slack, this isn't really in scope for an operator to help with. Objects are sync'd in unrelated reconcile calls so that's not an issue. However there's a very large chance that you should just eat the few minutes of downtime rather than trying to engineer a complex solution.

**Comment by tiagomeireles** [comment]
I understand that this is not in scope for the operator. I'm hoping to hear how others are handing this as i'm sure i'm not solely in this position. User facing downtime from a prometheus-operator upgrade is not acceptable for our use case.

**Comment by coderanger** [comment]
The bigger issue is that we shouldn't be doing things that require changing immutable fields, normal upgrades shouldn't touch any of those and that is probably a bug. If it's something that needed changing then there may not be much we can do but try and avoid it in the future.

**Comment by brancz** [comment]
We try our best never to do this, but it can happen from time to time, and there is not really anything else we can do much when it's necessary. We could have a setting where you are in control of doing the deletion instead, so at least you can do it on your own schedule. Do you think that would be helpful?

Also another thing we could explore (and I have no idea if this could actually be feasible but I think worth an exploration) could be to delete the StatefulSet in a non-cascading style, and then delete the pods one by one. I don't really like having this in the operator as arguably Kubernetes should be able to do something like that. But at the very least if this works, we can document it in combination with the first paragraph.

Thoughts?

**Comment by juliantaylor** [comment]
A non-cascading delete could be done, we do that to resize persistentvolumes attached to prometheus instances (in this case not operator managed).
Our script to do this looks about like this:

```
uid=$(kubectl_wrapper get statefulset "$sts" -o jsonpath='{.metadata.uid}')
pods=$(kubectl_wrapper get pods -o json | jq -r '.items[] | select(.metadata.ownerReferences[].uid=="'"$uid"'").metadata.name')

for pod in $pods; do
    log "Updating pvcs of pod $pod"
    changes=no
    while read -r name size; do
        rv=$(kubectl_wrapper get pvc "$name-$pod" -o jsonpath='{.metadata.resourceVersion}')
        out=$(kubectl_wrapper patch pvc "$name-$pod" -p '{"spec":{"resources": {"requests": {"storage":"'"$size"'"}}}}' -o jsonpath='{.metadata.resourceVersion}')
        if [[ $rv != "$out" ]]; then
            log "patched pvc $name-$pod to size $size"
            changes=yes
        fi
    done < <(echo "$statefulset_manifest" | jq -r '.spec.volumeClaimTemplates[] | .metadata.name + " " + .spec.resources.requests.storage')
    if [[ $changes != yes ]]; then
        die "No changes to be applied"
    fi  
done
# remove the statefulset so we can "update" the immutable size field
# will not restart the pods
echo "$statefulset_manifest" | kubectl_wrapper delete --cascade=false  -f -
echo "$statefulset_manifest" | kubectl_wrapper apply -f -

log "Restarting statefulset to trigger resize"
echo "$statefulset_manifest" | kubectl_wrapper rollout restart -f -
echo "$statefulset_manifest" | kubectl_wrapper rollout status -w --timeout=600s -f -

```

**Comment by brancz** [comment]
> Handling all cases in the operator likely becomes complex and error prone.

Agreed. This is neat though, thanks for sharing! We already have a `paused` option that makes Prometheus Operator ignore the Prometheus statefulset reconciliation, so this could be a solution as well.

**Comment by tiagomeireles** [comment]
I really like the non-cascading delete idea. Would the pod deletion be manual?

This is far simpler than the other options I have considered.

**Comment by brancz** [comment]
Yeah I think we would need pod deletion to be manual, but we can document something like @juliantaylor's comment.

We do still need the ability to disable the "automatic" conflict resolution we have today, and a mechanism to tell users that they must take action (I suppose an metric+alert and kubernetes event would do?)

**Comment by simonpasquier** [comment]
isn't the issue similar #2753 and the fact that the operator uses `podManagementPolicy: Parallel` to prevent statefulset reconciliation hanging forever?

**Comment by brancz** [comment]
I'm not 100% sure about that. Parallel pod management policy is just about the statefulset _potentially_ reconciling multiple pods at once, in that case we just need to limit the amount of pods it actually touches at once.

**Comment by tiagomeireles** [comment]
#2753 is a separate issue. This issue is regarding the operator resolving illegal updates to prometheus statefulsets causing a complete outage for prometheus setups that take multiple minutes to start. 

The "resolving illegal updates" process deletes the statefulset then recreates it. During this time all of the prometheus pods are unavailable to process queries. In our case prometheus takes 5+ minutes to become available.


I experimented with non-cascading delete and ran into some issues. 

I tested this process:
1. Reduce the prometheus-operator replica count to zero.
    - To prevent the older version of the prometheus-operator from recreating the statefulset.
2. Delete the prometheus statefulset with `--cascade=false`.
3. Upgrade prometheus-operator.
    - Creates the new statefulset.

The prometheus-config-reloader `Error`s and enters a `CrashLoopBackOff` with the following error `expand environment variables: found refrences to unset environment variable "SHARD"`.
`kubectl rollout restart sts prometheus` doesn't work. I believe it is because the statefulset selectors have changed.
Deleting pods one by one expecting the statefulset to recreate them doesn't work either. I have to delete all the pods before the statefulset will create new ones.

**Comment by brancz** [comment]
Thanks for testing! I would have tried something slightly different, although it's potentially the same result, but I think it's worth testing. For the sake of causing an illegal update, we could modify the requested storage space by:

1) Set `paused: true` in the `Prometheus` object.
2) Delete the statefulset with `--cascade=false`.
3) Set `paused: false` and modifying the storage resource request
4) Delete each pod individually by regex rather than `kubectl rollout`, as I believe you're right, rollout doesn't work in your case because the selector changed

Either way it seems it would be useful to at least allow administrators to opt-out of the "automatic conflict resolution".

**Comment by tiagomeireles** [comment]
@brancz Thank you for the `paused: true` tip. It improved the situation but doesn't resolve this problem.

```
Deleting pods one by one expecting the statefulset to recreate them doesn't work either. I have to delete all the pods before the statefulset will create new ones.
```


To work around user facing downtime I am currently running an additional prometheus object with the same scrape targets. My upgrades look like this:

1. Pause the additional prometheus object.
2. Upgrade prometheus-operator. 
3. Unpause the additional prometheus object after approximately 10m. 

I end up with a brief period where user may see a gap in graphs until the additional prometheus comes back online but its not a huge issue since we have caching via Thanos Query Frontend. 


Recently the upgrade behavior has IMO gotten worse. The prometheus statefulset objects now have a [version label](https://github.com/prometheus-operator/prometheus-operator/pull/3841/files#diff-930519027f29e25f016e4d8134774e70edcc0c7ab725c5a664dd66da55d07392R621). i.e `app.kubernetes.io/version=2.27.1` which now causes prometheus-operator to resolve conflicts for prometheus version upgrades.

**Comment by snuggie12** [comment]
@tiagomeireles I was able to use pause and slightly different steps with no issues. Here is the process I used. I'm on k8s 1.22 on GKE. If I'm vague on some steps it's because we use argocd/gitops so I won't have patch or edit commands listed:

1. Set `Prometheus` custom resource to `paused: true`
2. `kubectl get sts foo -o yaml > /tmp/sts.yaml`
3. edit `/tmp/sts.yaml` with new size. Only the new size should be changed and do *NOT* use a tool like `kubectl neat`. This resulted in parallel terminations so I'm guessing you need some `metadata` fields despite the fact my `uid`, `generation`, and `resourceVersion` all changed later.
4. `kubectl delete sts foo --cascade=orphan` if doing a `kubectl get pods -w` you'll see the pod names get repeated. I didn't bother to check but this is likely `.metadata.ownerReferences` being updated to `null`.
5. `kubectl edit pvc foo-n`. Edit only `.spec.resources.requests.storage` and not the part under `.status`
6. repeat until you get to `foo-0` PVC
7. at this point the PVCs should say the pods need restarted to take effect
8. `kubectl apply -f /tmp/sts.yaml`. You should see the same update happen on your pod watching. Likely it is adding the previously mentioned field back and taking ownership.
9. set `paused: false` and also update the size of the storage on the `Prometheus` custom resource
10. confirm nothing has gone wrong with the pods
11. `kubectl delete pod foo-n`
12. confirm pod is healthy
13. repeat until `foo-0` `Pod` is confirmed healthy

I do think the ability to not use `parallel` would be helpful so the last step is `kubectl rollout restart` but you're already doing a manual set of steps so it's not that big of a deal. I have two replicas and one thing I still can't figure out is why `foo-0`'s PVC says it is bound before I delete the pod but with an already healthy `foo-1` it wasn't a big deal to play it safe.

**Comment by FUSAKLA** [comment]
Hi, we bumped into this also, and we use in such cases the `kubectl delete sts --cascade=orphan` which leaves the pods running, then recreate the modifies sts, and then it automatically picks up the already existing pods and performs rollout of the changes. 

I believe this could be quite easy to do with the logic that is already there, where sts is deleted anyway. 

Would that be possible, and is there anything blocking this? If not and would be acceptable, I'd give it a shot.

**Comment by simonpasquier** [comment]
@FUSAKLA can you provide more details why the operator had to recreate the statefulset?
There's often a good reason for fields being immutable (https://github.com/kubernetes/kubernetes/issues/90519 is one example) and I'd be wary to introduce complex logic in the operator to handle all edge cases.
Having said that, it could be a good option to tell the operator "don't delete the statefulset if you can't update it" so you can handle the issue manually.

**Comment by sebhoss** [comment]
This might not be related to this ticket but the combined release of the prometheus-operator and prometheus-config-reloader image causes a recreation of every statefulset whenever an update for the prometheus-operator is applied because kubernetes will detect a change in those statefulset and recreate pods. If those images would be released independently, updates of the operator would not always cause recreation of statefulsets (assuming the config-reloader was not updated at the same time).

**Comment by simonpasquier** [comment]
@sebhoss the statefulset itself shouldn't be created in this case. It will get updated and the pods rolled out in sequence.
And it's less error prone to use the same tag for the operator and the config reloader.

**Comment by sebhoss** [comment]
@simonpasquier yep sorry I meant that the pods in the statefulset will be recreated and that causes a downtime if there is only a single replica

**Comment by simonpasquier** [comment]
@sebhoss the same is true if the pod has to be evicted from a pod. Basically in a single replica setup, many changes can disrupt the availability of Prometheus.

**Comment by sebhoss** [comment]
@simonpasquier yeah that's true, however most of our downtimes for single replica prometheus instances are caused by applying operator/config-reloader updates. It's not a big deal, but it has been noticed which caused me to spend some time today to dig into this :sweat_smile:

**Comment by zbialik** [comment]
A really common use-case for this is handling an increase in the PVC storage request. We will manually patch the PVC but want the StatefulSet definition to match what we patch. It'd be nice for the operator to do a `--cascade=orphan` in this scenario.
